applications/services/gui/modules/number_input.c

```diff
@@ -0,0 +1,493 @@
+    if(model->selected_column < number_input_get_row_size(model->selected_row) - 1) {
+        model->selected_column++;
+    } else {
+        model->selected_column = 0;
+    }
+    const NumberInputKey* keys = number_input_get_row(model->selected_row);
+    if(keys[model->selected_column].text == sign_symbol && !model->useSign) {
+        model->selected_column++;
+    }
+}
+
+static void prevent_to_large_number(NumberInputModel* model) {
+    if(strtol(furi_string_get_cstr(model->text_buffer), NULL, 10) > model->max_value) {
+        char* str = int32_to_string(model->max_value);
+        furi_string_set_str(model->text_buffer, str);
+        free(str);
+    }
+    // Added in prevent large, as it would block the input of small positive numbers
+    if(model->sign == '-' &&
+       strtol(furi_string_get_cstr(model->text_buffer), NULL, 10) < model->min_value) {
+        char* str = int32_to_string(model->min_value);
+        furi_string_set_str(model->text_buffer, str);
+        free(str);
+    }
+}
+
+static void prevent_to_small_number(NumberInputModel* model) {
+    if(strtol(furi_string_get_cstr(model->text_buffer), NULL, 10) < model->min_value) {
+        char* str = int32_to_string(model->min_value);
+        furi_string_set_str(model->text_buffer, str);
+        free(str);
+    }
+}
+
+static void number_input_sign(NumberInputModel* model) {
+    if(model->sign == '-') {
+        model->sign = '+';
+    } else {
+        model->sign = '-';
+    }
+    int32_t number = strtol(furi_string_get_cstr(model->text_buffer), NULL, 10);
+    number = number * -1;
+    char* str = int32_to_string(number);
+    furi_string_set_str(model->text_buffer, str);
+    free(str);
+    prevent_to_small_number(model);
+    prevent_to_large_number(model);
+}
+
+/** Handle OK button
+ *
+ * @param      model  The model
+ */
+static void number_input_handle_ok(NumberInputModel* model) {
+    char selected = number_input_get_row(model->selected_row)[model->selected_column].text;
+    char temp_str[2];
+    temp_str[0] = selected;
+    temp_str[1] = '\0';
+    if(selected == enter_symbol) {
+        prevent_to_small_number(model);
+        prevent_to_large_number(model);
+        model->callback(model->callback_context);
+    } else if(selected == backspace_symbol) {
+        number_input_backspace_cb(model);
+    } else if(selected == sign_symbol) {
+        number_input_sign(model);
+    } else {
+        furi_string_cat_str(model->text_buffer, temp_str);
+        prevent_to_large_number(model);
+    }
+    model->clear_default_text = false;
+}
+
+/** Draw callback
+ *
+ * @param      canvas  The canvas
+ * @param      _model  The model
+ */
+static void number_input_view_draw_callback(Canvas* canvas, void* _model) {
+    NumberInputModel* model = _model;
+    uint8_t text_length = model->text_buffer ? furi_string_utf8_length(model->text_buffer) : 0;
+    UNUSED(text_length);
+
+    canvas_clear(canvas);
```

Comment:
clear is done by canvas before each draw call. there is no need clear in your draw call callback